Accept initializers for weights and biases in NStepLSTM

niboshi opened the pull request:
This PR is to allow arbitrary initializers in `NStepLSTM.__init__` (See issue #1816).

unnonouno commented:
Thank you, but @aonotas is currently working for refactoring n-step-lstm. It is better to merge this PR after @aonotas 's work will be finished.

niboshi commented:
OK, I got it.

niboshi commented:
Rebased to the current master.

Note that `initializers.init_weight()` defaults to the scaled normal distribution when the second argument (`initial_bias`) is `None`.

unnonouno commented:
Merge the master. Many __init__ method for Function are fixed.

niboshi commented:
Rebased to master.

unnonouno commented on chainer/links/connection/n_step_lstm.py:
```diff
@@ -53,12 +61,12 @@ def __init__(self, n_layers, in_size, out_size, dropout, use_bi_direction,
-                        setattr(weight, 'b%d' % j, b)
+                        name_w = 'w{}'.format(j)
+                        name_b = 'b{}'.format(j)
+                        w = variable.Parameter(initialW, (out_size, w_in))
```
Do we need to use `initializers._get_initializer` for `initialW`?

niboshi commented:
Fixed to call `_get_initializer` for `initialW`. Also updated docstring.

I understand the only point of `_get_initializer` is to convert `None` to the "default" initializer, which is `LeCunNormal`. Is that correct?

I think it could be done in better way.
If this understanding is correct, I will make another PR to refactor.

unnonouno commented:
Yes, and it is the same behavior as links.Linear.

unnonouno commented:
LGTM

MannyKayy commented:
This PR was overwritten during the [rnn code refactor].

(https://github.com/chainer/chainer/commit/f8e99052b84375faf86bce88c2b86b57db1f6885#diff-38391e4f2027e281d76ad22b877bd019).

![image](https://user-images.githubusercontent.com/7952854/39865715-b23fcaf6-5445-11e8-8060-f50848978401.png)


Runtime Info:

Chainer: 5.0.0a1
NumPy: 1.13.0
CuPy:
  CuPy Version          : 5.0.0a1
  CUDA Root             : /usr/local/cuda
  CUDA Build Version    : 8000
  CUDA Driver Version   : 9020
  CUDA Runtime Version  : 8000
  cuDNN Build Version   : 7102
  cuDNN Version         : 7102
  NCCL Build Version    : 2104

delta2323 commented:
Thank you very much for your report. What you pointed out seems to be correct and should be considered as a bug.

delta2323 commented:
I made an issue dealing with it (#4755). By the way, this PR was merged in v3.0.0rc1, but is not written in [the release note](https://github.com/chainer/chainer/releases/tag/v3.0.0rc1) of this version. I think we add this PR to it.
